### DISCUSSION TITLE
How can I reference a endpoint in my client app from a custom ExecutableResouce?

### DISCUSSION THREAD

**Discussion by JobaDiniz** [microsoft/aspire#7002] [Q&A] [answered]
I'm new to creating custom resources in Aspire. Not even sure whether the below code is correct way of doing it.
But essentially I'm trying to create a `ExecutableResource` that launches `temporal` CLI (`temporal server start-dev`). This command creates 3 endpoints:

```
PS C:\Users\jdiniz> temporal server start-dev
CLI 1.1.2 (Server 1.25.2, UI 2.32.0)

Server:  localhost:7233
UI:      http://localhost:8233
Metrics: http://localhost:61644/metrics
```

```csharp
public class TemporalExecutableResource(string name, string? executablePath) :
    ExecutableResource(name, executablePath ?? "temporal", string.Empty), IResourceWithEndpoints{}

public static class TemporalBuilderExtensions
{
    public static IResourceBuilder<TemporalExecutableResource> AddTemporalDevelopmentServer(this IDistributedApplicationBuilder builder, string name, string? executablePath = null, string[]? args = null)
    {
        args = args ?? [];
        string[] allArgs = args is { Length: > 0 }
            ? ["server", "start-dev", .. args]
            : ["server", "start-dev"];
        var temporal = new TemporalExecutableResource(name, executablePath);
        return builder.AddResource(temporal)
            .WithArgs(allArgs)
            .WithEndpoint(port: 7233, scheme: "grpc", name: "server", isProxied: false, isExternal: true)
            .WithHttpEndpoint(port: 8233, name: "ui", isProxied: false)
            .WithHttpEndpoint(port: 52723, name: "metrics", isProxied: false);
    }
}
```

This does not work as it should, but I'll figure out later, my question is more *basic*: 

How can I **reference** these endpoints in the client app?

```csharp
options.ClientOptions = new Temporalio.Client.TemporalClientConnectOptions("HOW TO REFERENCE the SERVER ENDPOINT HERE?")
{
    Namespace = "default",
    DataConverter = DataConverter.Default with { PayloadConverter = new DefaultPayloadConverter(NetFlowJsonSerializerOptions.Definition) }
};
```

**Comment by davidfowl** [comment, marked as answer]
First, there's a working temporal hosting integration on NuGet:

NuGet - https://www.nuget.org/packages/InfinityFlow.Aspire.Temporal
Github - https://github.com/InfinityFlowApp/aspire-temporal

> How can I reference these endpoints in the client app?

Have you read 
- https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/app-host-overview?tabs=docker#connection-string-and-endpoint-references
- https://learn.microsoft.com/en-us/dotnet/aspire/extensibility/custom-hosting-integration
- https://learn.microsoft.com/en-us/dotnet/aspire/extensibility/custom-client-integration

Most integrations implement `IResourceWithConnectionString` and create a connection string expression that will get passed to the "cleint" or "consuming" resource when `WithReference` is used. 

There're extensive comments on what these APIs do and what they inject into the target application:

https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.resourcebuilderextensions.withreference?view=dotnet-aspire-8.0

ExecutableResource already implemets IResourceWithEndpoints, so adding it to your resouce implementation is not useful. At a lower level, you can get an endpoint by name by calling [GetEndpoint](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.resourcebuilderextensions.getendpoint?view=dotnet-aspire-8.0) and you can pass it to another resource (this example shows passing it to a project resource):

```C#
var builder = DistributedApplication.CreateBuilder(args);

var temporal = builder.AddTemporalDevelopmentServer("temporal");

builder.AddProject<Projects.Api>("api")
    .WithReference(temporal.GetEndpoint("grpc"));

builder.Build().Run();
```

This will pass the resolved URL (at the appropriate time) to the target project using the service discovery format described [here](https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/app-host-overview?tabs=docker#service-endpoint-environment-variable-format).

OR you can [set environment variables directly](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.resourcebuilderextensions.withenvironment?view=dotnet-aspire-8.0#aspire-hosting-resourcebuilderextensions-withenvironment-1(aspire-hosting-applicationmodel-iresourcebuilder((-0))-system-string-aspire-hosting-applicationmodel-endpointreference)):

```C#
var builder = DistributedApplication.CreateBuilder(args);

var temporal = builder.AddTemporalDevelopmentServer("temporal");

builder.AddProject<Projects.Api>("api")
    .WithEnvironment("TemporalDevUrl", temporal.GetEndpoint("grpc"));

builder.Build().Run();
```

**Reply by JobaDiniz** [reply]
thanks. I'm starting to get the _hang of it_ :). The missing piece was to add a `IResourceWithConnectionString`. But as you said, I could chose to not implement the interface and use `.WithReference(temporal.GetEndpoint("grpc")`. *Did not know about this*.

I've implemented like this:

```csharp
public class TemporalExecutableResource(string name, string? executablePath) :
    ExecutableResource(name, executablePath ?? "temporal", string.Empty), IResourceWithConnectionString
{
    internal const string ServerEndpointName = "server";

    private EndpointReference? serverReference;
    public EndpointReference ServerEndpoint =>
        serverReference ??= new(this, ServerEndpointName);

    public ReferenceExpression ConnectionStringExpression =>
        ReferenceExpression.Create(
            $"{ServerEndpoint.Property(EndpointProperty.Host)}:{ServerEndpoint.Property(EndpointProperty.Port)}"
        );
}

public static class TemporalBuilderExtensions
{
    public static IResourceBuilder<TemporalExecutableResource> AddTemporalDevelopmentServer(this IDistributedApplicationBuilder builder, string name, string? executablePath = null, string[]? args = null)
    {
        args = args ?? [];
        string[] allArgs = args is { Length: > 0 }
            ? ["server", "start-dev", .. args]
            : ["server", "start-dev"];
        var temporal = new TemporalExecutableResource(name, executablePath);
        return builder.AddResource(temporal)
            .WithArgs(allArgs)
            .WithEndpoint(port: 7233, scheme: string.Empty, name: TemporalExecutableResource.ServerEndpointName, isProxied: false)
            .WithHttpEndpoint(port: 8233, name: "ui", isProxied: false)
            .WithHttpEndpoint(port: 52723, name: "metrics", isProxied: false);
    }
}
```

**Reply by davidfowl** [reply]
Great! I opened a [doc issue](https://github.com/dotnet/docs-aspire/issues/2340) to write up how endpoint references work.
